zipline/pipeline/factors/factor.py

```diff
@@ -619,7 +626,21 @@ def rank(self, method='ordinal', ascending=True, mask=NotSpecified):
         :func:`scipy.stats.rankdata`
         :class:`zipline.pipeline.factors.factor.Rank`
         """
-        return Rank(self, method=method, ascending=ascending, mask=mask)
+
+        if groupby is NotSpecified:
+            return Rank(self, method=method, ascending=ascending, mask=mask)
+
+        else:
+            def rank(row):
+                return rankdata(row if ascending else -row, method=method)
```

Comment:
Noticed your question re: this not being a local got clobbered.

From a correctness perspective, it shouldn't matter, and from a performance perspective, the difference between loading a local variable and loading a closure variable is small (compared to loading a global, for example, which is a dict lookup).  The main cost of this strategy is that we're branching on every row even though we only need to do the check once.  If we want to be really tight here, I would probably write this as:

```
from functools import partial

if groupby is NotSpecified:
    return Rank(...)
elif ascending:
    rank = partial(rankdata, method=method)
else:
    def rank(row):
        return rankdata(-row, method=method)    
```

Another *possible* speedup would be to pass `-self` as `factor` instead of negating the row each time.  Naively, that should consume less CPU, since we can do the negation vectorized on a larger array, but that also means that we'll temporarily have a full copy of the (negated) array in memory at, which might actually degrade performance by requiring more allocations.  I'd have to profile to see if one were better than the other.